### PR TITLE
Story1

### PR DESCRIPTION
**Pull request by melody-sylvestre**

### CODE REVIEW

File: src/dbFunctions.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+function connectToDb(string $dbname): PDO {
+    $db = new PDO("mysql:host=db; dbname=$dbname", "root", "password");
+    $db -> setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+    
+    return $db;  
+}
+
+function getItemsFromDb(PDO $db): array {
+    $query = $db->prepare("SELECT * FROM yarn");
```
Comment (norayna) [inline src/dbFunctions.php]: according to best practice the table should be called `yarns`

File: src/test/collectionHandlingFunctions.php [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        $yarnExample = ["name" => NULL, 
+                        "shade" => "Rainbow", 
+                        "dominantColour" => "purple",
+                        "composition" => "Cashmere",
+                        "yarnType" => "Super Chunky", 
+                        "lengthInMeters" => 100,
+                        "skeinNumber" => 2, 
+                        "project"=> "Cool rainbow hat",
+                        "image" => "blue_scarf.jpg"      
+                    ];
```
Comment (norayna) [inline src/test/collectionHandlingFunctions.php]: it is correct to put [ on line 16, then everything starting from the line below, the indentation should be different as well (make sure to change this in all the test functions)

File: src/test/collectionHandlingFunctions.php [hunk 3]
```diff
@@ -0,0 +1,314 @@
+<?php
+require "../collectionHandlingFunctions.php";
+use PHPUnit\Framework\TestCase;
+
+class collectionHandlingFunctions extends TestCase 
+{
+
+    public function testFailureDescribeCollectionItem_EmptyItem()
```
Comment (norayna) [inline src/test/collectionHandlingFunctions.php]: should be `_emptyItem` as opposed to `_EmptyItem` for consistency with the rest of the code

File: src/test/collectionHandlingFunctions.php [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        $expectedOutput .= '</div>'; 
+        
+        $actualOutput = describeCollectionItem($yarnExample);
+
+        $this->assertEquals($expectedOutput, $actualOutput);
+    }
+
+ //The following functions check that the output is correct when one of the fields is missing 
+ // that item in the database    
+    public function testSuccessDescribeCollectionItem_undefinedShade()
```
Comment (norayna) [inline src/test/collectionHandlingFunctions.php]: the comments should be in line with the function indentation wise

File: src/test/collectionHandlingFunctions.php [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+        $expectedOutput .= '<li>Available for a new project!</li>';
+        $expectedOutput .= '</ul>';
+        $expectedOutput .= '</div>'; 
+        
+        $actualOutput = describeCollectionItem($yarnExample);
+
+        $this->assertEquals($expectedOutput, $actualOutput);
+    }
+
+// in the following functions, we check that the bullet point of the missing property is not printed
```
Comment (norayna) [inline src/test/collectionHandlingFunctions.php]: isn't this comment basically the same as the one on line 61?

File: styles/style.css [hunk 6]
```diff
@@ -0,0 +1,43 @@
+html {
+    font-size: 10px; 
+    font-family: 'Didact Gothic', Tahoma, sans-serif;
```
Comment (norayna) [inline styles/style.css]: i think you need to import this font in order to use it

File: styles/style.css [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+h4 {
+    font-size: 2rem;
+}
+
+li {
+    font-size: 2rem;
+}
+
+h1, h2 {
+    margin: 50px 50px;
```
Comment (norayna) [inline styles/style.css]: this can be shortened to `margin: 50px;`

File: src/collectionHandlingFunctions.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+function describeCollectionItem(array $item): string {
+    
+    if($item===[]){
```
Comment (ashleycoles) [inline src/collectionHandlingFunctions.php]: Being nitpicky, there should be a space between the if and the (, and the ) and the {

`if ($item === []) {`

File: src/collectionHandlingFunctions.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    
+    } else if(!isset($item["name"])) {
+        throw new Exception("Undefined yarn name - there must be something wrong with this item...\n");  
+    
+    } else {
+        $description = '<div class="collection_item">';
+        
+        if(isset($item["name"]) && isset($item["shade"])){
+            $description .= '<h3 tabindex="0">' . $item["name"]. " " . $item["shade"] . '</h3>';
+
```
Comment (ashleycoles) [inline src/collectionHandlingFunctions.php]: Remove the empty line here

File: src/dbFunctions.php (hunk 1 above)
Comment (ashleycoles) [reply src/dbFunctions.php]: I would also say you shouldn't do use the * here - it's ok when quickly testing something, but it's better to specifically list the fields you want to select, even if it's all of them.

The reasoning is that the table might expand in the future to contain more fields that you may not want to retrieve in this particular query.

File: src/test/collectionHandlingFunctions.php [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+    public function testFailureDescribeCollectionItem_EmptyItem() 
+    {
+        $yarnExample = [];
+        $this->expectException(Exception::class); 
+        describeCollectionItem($yarnExample);
+    }
+    
+    public function testFailureDescribeCollectionItem_undefinedName() {
+        $yarnExample = ["name" => NULL,
```
Comment (ashleycoles) [inline src/test/collectionHandlingFunctions.php]: You should put the first array item on the next line after the [, indented by one level.

Here's an example of what I mean

```
$example = [
    "foo" => "bar",
    "bar" => "baz"
];
```

Comment (ashleycoles) [reply src/test/collectionHandlingFunctions.php]: This applies to the other arrays in this test file

**Review by ashleycoles — APPROVED** [review]
A few comments from you (plus some from Rayna) - really nice work!

File: src/collectionHandlingFunctions.php (hunk 8 above)
Comment (melody-sylvestre) [reply src/collectionHandlingFunctions.php]: done! (I also changed the other if statements)

File: src/collectionHandlingFunctions.php (hunk 9 above)
Comment (melody-sylvestre) [reply src/collectionHandlingFunctions.php]: I removed the empty line at line 16

**Review by ashleycoles — CHANGES_REQUESTED** [review]
Whoooops I accidentally hit approve on my PR when I meant to select request changes!

File: src/test/collectionHandlingFunctions.php (hunk 10 above)
Comment (melody-sylvestre) [reply src/test/collectionHandlingFunctions.php]: done!

File: src/dbFunctions.php (hunk 1 above)
Comment (melody-sylvestre) [reply src/dbFunctions.php]: Done: I changed the name of my table in the database and edited the code.

File: src/test/collectionHandlingFunctions.php (hunk 2 above)
Comment (melody-sylvestre) [reply src/test/collectionHandlingFunctions.php]: done!

File: src/test/collectionHandlingFunctions.php (hunk 3 above)
Comment (melody-sylvestre) [reply src/test/collectionHandlingFunctions.php]: Done!

File: src/test/collectionHandlingFunctions.php (hunk 4 above)
Comment (melody-sylvestre) [reply src/test/collectionHandlingFunctions.php]: Done

File: src/test/collectionHandlingFunctions.php (hunk 5 above)
Comment (melody-sylvestre) [reply src/test/collectionHandlingFunctions.php]: Not totally: for these functions, I check that a missing field results in a missing bullet point,  whereas in the previous functions I would change the some other parts of the text (e.g. the content of `` tag). But maybe this comment is not that useful? I removed it.

File: styles/style.css (hunk 6 above)
Comment (melody-sylvestre) [reply styles/style.css]: I removed it

File: styles/style.css (hunk 7 above)
Comment (melody-sylvestre) [reply styles/style.css]: done

**Review by ashleycoles — APPROVED** [review]
Perfect!

**Pull request merged**